### PR TITLE
Socrata Metadata Publishing

### PR DESCRIPTION
**Pull request by Charlie-Henry**
Moving our last script from [atd-executive-dashboard](https://github.com/cityofaustin/atd-executive-dashboard) to its own repo. 

Testing if you're interested, clone this repo and checkout this branch, you do not need to be on the VPN:
1. Fill in env_file_template with the secrets in "Socrata Key ID, Secret, and Token" in 1pass.
2. `$ docker build . -t dts-socrata-reporting:production`
3. `$ docker run -it --env-file env_file dts-socrata-reporting:production /bin/bash` 
4. `$ python etl/socrata_metadata_pub.py`

This might take a while but check for errors and check if it completes 👍 

Commit history might be a disaster as I accidentally pushed to main 🤦‍♂️

### CODE REVIEW

**Comment by chiaberry** [comment]
it took a long time to run and I got
```
2024-05-27 21:15:27,256 INFO: Transforming data
Traceback (most recent call last):
  File "/app/etl/socrata_metadata_pub.py", line 210, in <module>
    main()
  File "/app/etl/socrata_metadata_pub.py", line 204, in main
    data = transform(data)
  File "/app/etl/socrata_metadata_pub.py", line 186, in transform
    filtered_data["row_count"] = row_count[0]["count"]
KeyError: 0
```
I am going to run it again

**Review by chiaberry — APPROVED** [review]
It worked on the subsequent try. I dunno why it errored the first time

**Review by frankhereford — APPROVED** [review]
Thanks Charlie. Ran great, 13 minutes run time.

```
root@2f09c8b93b2d:/app# time python ./etl/socrata_metadata_pub.py 
2024-05-28 12:20:04,671 INFO: Extracting data
2024-05-28 12:20:12,648 INFO: Transforming data
2024-05-28 12:33:13,272 INFO: Publishing data to Socrata
2024-05-28 12:33:14,702 INFO: {'Errors': 0, 'Rows Created': 316, 'Rows Updated': 0, 'Rows Deleted': 0}

real    13m10.248s
user    0m8.483s
sys     0m0.535s
root@2f09c8b93b2d:/app# 
```

File: etl/socrata_metadata_pub.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    return data
+
+
+def convert_tz(date_str):
+    if not date_str:
+        return None
+
+    # Get UTC timestamp as datetime object
+    utc_datetime = datetime.strptime(date_str, "%Y-%m-%dT%H:%M:%S.%fZ")
+    utc_timezone = pytz.timezone("UTC")
```
Comment (johnclary) [inline etl/socrata_metadata_pub.py:98]: not necessary at all to revisit this but wanted to mention that i'm excited about the new [ZoneInfo](https://docs.python.org/3/library/zoneinfo.html) library that finally makes it possible to do this with built-ins in python v3.9+

**Review by johnclary — APPROVED** [review]
have not tested: still looks great! would be cool to make this use `asyncio` just for funzees

🚢 🙏

**Comment by Charlie-Henry** [comment]
I added async for fetching the dataset row count. It would be good to get y'alls eyes on this again @johnclary et al. I was sometimes getting runtime down to about a minute and other times it would be significantly slower but that might just be Socrata rate limiting our API key?

File: requirements.txt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+sodapy==2.1.*
+pytz~=2023.3.post1
+asyncio==3.4.*
+aiohttp==3.9.*
+requests==2.31.*
```
Comment (frankhereford) [inline requirements.txt]: Can you put a new line here at the EOF for posix compliance?

**Review by frankhereford — APPROVED** [review]
Looks like you sped it up considerably:

```
root@0510111a4acb:/app# time python etl/socrata_metadata_pub.py 
2024-06-14 17:26:13,375 INFO: Extracting data
2024-06-14 17:26:21,797 INFO: Transforming data
2024-06-14 17:32:30,188 INFO: Publishing data to Socrata
2024-06-14 17:32:31,198 INFO: {'Errors': 0, 'Rows Created': 319, 'Rows Updated': 0, 'Rows Deleted': 0}

real    6m17.934s
user    0m3.108s
sys     0m0.239s
root@0510111a4acb:/app# 
```

Do you mind tossing a new line in that requirements.txt?

**Pull request merged**